govcd/vapp_network_test.go

```diff
@@ -142,3 +149,119 @@ func (vcd *TestVCD) Test_GetVappNetworkByNameOrId(check *C) {
+		&types.NatRule{VMRule: &types.NatVMRule{
+			ExternalPort: 80, InternalPort: 22, VMNicID: 0,
+			VAppScopedVMID: vm2.VM.VAppScopedLocalID, Protocol: "UDP"}}},
+		"portForwarding", "allowTraffic")
+	check.Assert(err, IsNil)
+	check.Assert(result, NotNil)
+	check.Assert(len(result.Configuration.Features.NatService.NatRule), Equals, 2)
+
+	// verify
+	check.Assert(result.Configuration.Features.NatService.Policy, Equals, "allowTraffic")
+	check.Assert(result.Configuration.Features.NatService.NatType, Equals, "portForwarding")
+
+	check.Assert(result.Configuration.Features.NatService.NatRule[0].VMRule.InternalPort, Equals, 22)
+	check.Assert(result.Configuration.Features.NatService.NatRule[0].VMRule.ExternalPort, Equals, -1)
+	check.Assert(result.Configuration.Features.NatService.NatRule[0].VMRule.VAppScopedVMID, Equals, vm.VM.VAppScopedLocalID)
+	check.Assert(result.Configuration.Features.NatService.NatRule[0].VMRule.VMNicID, Equals, 0)
+
+	check.Assert(result.Configuration.Features.NatService.NatRule[1].VMRule.InternalPort, Equals, 22)
+	check.Assert(result.Configuration.Features.NatService.NatRule[1].VMRule.ExternalPort, Equals, 80)
+	check.Assert(result.Configuration.Features.NatService.NatRule[1].VMRule.VAppScopedVMID, Equals, vm2.VM.VAppScopedLocalID)
+	check.Assert(result.Configuration.Features.NatService.NatRule[1].VMRule.VMNicID, Equals, 0)
+
+	result, err = vapp.UpdateNetworkNatRules(uuid, []*types.NatRule{&types.NatRule{OneToOneVMRule: &types.NatOneToOneVMRule{
+		MappingMode: "automatic", VMNicID: 0,
+		VAppScopedVMID: vm.VM.VAppScopedLocalID}},
+		&types.NatRule{OneToOneVMRule: &types.NatOneToOneVMRule{
+			MappingMode: "manual", VMNicID: 0,
+			VAppScopedVMID:    vm2.VM.VAppScopedLocalID,
+			ExternalIPAddress: takeStringPointer("192.168.100.1")}}},
+		"ipTranslation", "allowTrafficIn")
+	check.Assert(err, IsNil)
+	check.Assert(result, NotNil)
+	check.Assert(len(result.Configuration.Features.NatService.NatRule), Equals, 2)
+
+	// verify
+	check.Assert(result.Configuration.Features.NatService.Policy, Equals, "allowTrafficIn")
+	check.Assert(result.Configuration.Features.NatService.NatType, Equals, "ipTranslation")
+
+	check.Assert(result.Configuration.Features.NatService.NatRule[0].OneToOneVMRule.MappingMode, Equals, "automatic")
+	check.Assert(result.Configuration.Features.NatService.NatRule[0].OneToOneVMRule.VAppScopedVMID, Equals, vm.VM.VAppScopedLocalID)
+	check.Assert(result.Configuration.Features.NatService.NatRule[0].OneToOneVMRule.VMNicID, Equals, 0)
+
+	check.Assert(result.Configuration.Features.NatService.NatRule[1].OneToOneVMRule.MappingMode, Equals, "manual")
+	check.Assert(result.Configuration.Features.NatService.NatRule[1].OneToOneVMRule.VAppScopedVMID, Equals, vm2.VM.VAppScopedLocalID)
+	check.Assert(result.Configuration.Features.NatService.NatRule[1].OneToOneVMRule.VMNicID, Equals, 0)
+	check.Assert(*result.Configuration.Features.NatService.NatRule[1].OneToOneVMRule.ExternalIPAddress, Equals, "192.168.100.1")
```

Comment:
We should check that the pointer is not nil before comparing its value

Reply:
added